Add project baseconfig

Luffy2004-c opened the pull request:
## Description

<!-- Add a more detailed description of the changes if needed. -->
Add project baseconfig

## Related Issue

<!-- If your PR refers to a related issue, link it here. -->
https://github.com/Undertone0809/gcop/issues/126

## Type of Change

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [x] 📚 Examples / docs / tutorials / dependencies update
- [x] 🔧 Bug fix (non-breaking change which fixes an issue)
- [] 🥂 Improvement (non-breaking change which improves an existing feature)
- [ ] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [ ] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/Undertone0809/gcop/blob/main/CODE_OF_CONDUCT.md) document.
- [ ] I've read the [`CONTRIBUTING.md`](https://github.com/Undertone0809/gcop/blob/main/CONTRIBUTING.md) guide.
- [x] I've updated the code style using `make codestyle`.
- [x] I've written tests for all new methods and classes that I created.
- [x] I've written the docstring in Google format for all the methods and classes that I used.

Luffy2004-c committed:
feat add project base config

Luffy2004-c committed:
optimize : add <dataclass> for local config

Luffy2004-c committed:
test: add test for init_config

Luffy2004-c committed:
feat : add auto commit learn from history

Luffy2004-c committed:
code format

Luffy2004-c committed:
optimise init-project command

Luffy2004-c committed:
optimise the way to get config

Luffy2004-c committed:
refactor : optimise get_config that is added to project configuration logic

Luffy2004-c committed:
fix: fix the logic of check_model_config()

Luffy2004-c committed:
add unit-test for get_config

Luffy2004-c committed:
fix bugs : fix the logic of getting config while project config is existing

Luffy2004-c committed:
refactor : code format

Luffy2004-c committed:
refactor ：improve get_config logic

Luffy2004-c committed:
docs ：Added an introduction to the init-project and show-config commands

Luffy2004-c committed:
docs : Added  docs about how to initialize a project-level configuration

Undertone0809 commented on docs/guide/commands.md:
```diff
@@ -131,4 +139,24 @@ Some advanced features like line count by language require additional tools (e.g
 The `git amend` command modifies Git history. Use it with caution, especially if you've already pushed the commit you're amending to a shared repository.
 :::
 
+7. To initialize the GCOP configuration in the project:
```
我理解这个不算第七点，算一个新的 example case。

Undertone0809 commented on docs/other/config-your-project-config.md:
```diff
@@ -0,0 +1,65 @@
+# Configure Your Project
```
可以改成 Project based configuration

Undertone0809 commented on gcop/__main__.py:
```diff
@@ -492,6 +513,35 @@ def commit_command(
+    logger.color_info(f"Current gcop config: {config}")
+
+
 @app.command(name="help")
```
help 需要加上对应的命令

Undertone0809 commented on gcop/__main__.py:
```diff
@@ -492,6 +513,35 @@ def commit_command(
+def show_config_command():
+    """command to show the current gcop config"""
+    config = get_config()
+    logger.color_info(f"Current gcop config: {config}")
```
这个输出要格式化一下

Luffy2004-c committed:
Update the project configuration document, restructure the logic for obtaining configuration, and optimize the command help information.

Luffy2004-c committed:
更新项目配置文档，重构获取配置逻辑，优化命令帮助信息

Luffy2004-c committed:
修复：在 GcopConfig 中使用 deepcopy 复制默认配置以避免意外修改

Luffy2004-c committed:
Fix: Use deepcopy in GcopConfig to copy the default configuration to avoid accidental modifications.

Luffy2004-c committed:
docs：更新配置文档，增加配置优先级说明及项目级配置创建指南

Luffy2004-c committed:
docs：Update the configuration document, add explanations on configuration priority and create guidelines for project-level configuration.

Undertone0809 commented on gcop/prompt.py:
commmit_message_history  的功能和正常的 commit_template 你觉得是互斥的吗？

个人感觉从效果上两种模式的互斥的，会相互干扰。

Undertone0809 replied:
需要讨论一下，得出一个结论，因为如果是互斥着，意味着这部分的逻辑和对应的单测需要更新一下 @Luffy2004-c

Undertone0809 replied:
先测一下效果

Undertone0809 commented:
- 还需要把主分支的更新 merge 进来，现在和主分支是有冲突的
- 需要把那些 commit message 改成英文，不要用中文的 commit message。

Undertone0809 commented on tests/test_get_config.py:
```diff
@@ -0,0 +1,11 @@
+from gcop.config import GcopConfig
+
+
+def test_get_config():
```
这里是不是应该 mock 一下 config 数据，比如 mock project based config 然后测试一下得到的 config 是不是 project based config

Undertone0809 commented on gcop/config.py:
```diff
@@ -74,60 +106,96 @@ class GcopConfig(metaclass=Singleton):
 
     """
 
+    default_config: ClassVar[Dict[str, Any]] = {
```
这个变量是不是应该单独拿出来作为常量。

Undertone0809 commented on gcop/config.py:
```diff
@@ -74,60 +106,96 @@ class GcopConfig(metaclass=Singleton):
+            logger.color_info(
+                "Warning:You are using the example configuration for the model."
+                "Please replace the example values with "
+                "your actual model_name ,api_key and api_base to ensure the project "  # noqa
```
可以再附上一个文档链接在里面，方便参考

Luffy2004-c committed:
Merge branch 'add-project-baseconfig' of https://github.com/Luffy2004-c/gcop into add-project-baseconfig

Luffy2004-c committed:
feat: Add the default configuration file and update the initialization logic to use the new configuration.

Luffy2004-c committed:
Merge branch 'main' into add-project-baseconfig

Luffy2004-c committed:
feat: Enhance model configuration initialization and add project config loading tests

Undertone0809 commented on gcop/config.py:
```diff
@@ -1,10 +1,34 @@
-from gcop.utils import get_default_storage_path, read_yaml
+from gcop.utils import Color, get_default_storage_path, logger, read_yaml
+
+with open("default_config.json") as f:
```
为什么不直接写在代码里呢？你这样的话每次读取脚本都要多做一步 io 操作。

Undertone0809 commented on gcop/config.py:
```diff
@@ -29,6 +53,18 @@ class ModelConfig:
 
+    def check_model_config(self) -> bool:
+        if (
+            self.model_name == "provider/name,eg openai/gpt-4o"
```
如果你 default config 是一个 py 变量的话，而不是 json 文件，这里就可以直接用 if self.model_name == DEFAULT_CONF.model_name

Undertone0809 committed:
refactor: improve configuration management structure

- Rename `default_config` to `DEFAULT_CONFIG` for consistency
- Remove JSON file handling and replace with direct dictionary definition
- Enhance documentation and structure of `GcopConfig` and `ModelConfig`
- Improve configuration loading and merging logic

These changes aim to make the configuration management system more maintainable and easier to understand, without changing its core functionality.

Undertone0809 committed:
refactor: remove default_config.json

- Delete default_config.json file
- Remove unused configuration settings

This change is part of a cleanup effort to remove outdated or unused configuration files, simplifying the project structure and reducing potential confusion.

zeelanddd committed:
Merge branch 'add-project-baseconfig' of https://github.com/luffy2004-c/gcop into add-project-baseconfig

Undertone0809 committed:
docs: update doc text

Undertone0809 commented:
@Luffy2004-c can use https://github.com/Undertone0809/conftier to refactor it.

Luffy2004-c committed:
feat: add configuration schema classes

- Implement ModelConfig class for AI model configuration
- Add GcopConfig class for overall application configuration
- Include validation method for model configuration

These classes provide a structured way to manage application
configuration with proper type hints and default values. The
ModelConfig includes validation to ensure example values are
not used in production.

Luffy2004-c committed:
refactor:Use conftier to reconfigure the related modules of gcop configuration

Luffy2004-c committed:
Update Python version requirements and dependencies

Undertone0809 commented on gcop/__main__.py:
```diff
@@ -493,6 +527,130 @@ def commit_command(
     actions[response]()
 
 
+# OPTIMIZE:create_project_config_template返回文件时，附带是否新建的属性，以给出粒度更细的用户提示  # noqa: E501
```
English plz

Undertone0809 commented on docs/guide/commands.md:
```diff
@@ -83,4 +87,85 @@ Display detailed information about the current git repository. This command prov
 - Version control information (latest tag, branch count, untracked files)
 - Advanced details (submodules, latest merge commit, file type statistics)
 
+### `gcop init-project`
```
Seems I have not found the detail use cases of project and user cases. You can create a new case of custom git commit message template when a team need to use.

It seems that I haven't found the detailed use cases of the [The advantages of project config] use cases yet. 

When the team needs to use it, you can create a new custom git commit message template. --> This case can be displayed in docs.

Undertone0809 commented on gcop/__main__.py:
```diff
@@ -493,6 +527,130 @@ def commit_command(
+        typer.echo("No project config found")
+
+
+@app.command(name="set-config")
```
Need to assert schema for every key.

Luffy2004-c committed:
feat: add config key validation

- Implement is_key_in_config function to check nested keys
- Validate keys in set_config command before processing
- Improve error handling for invalid config keys

This change prevents setting invalid configuration keys by
validating them against the config schema first. It handles
both dataclass attributes and dictionary keys for nested
config structures.

Undertone0809 committed:
Merge branch 'add-project-baseconfig' of https://github.com/luffy2004-c/gcop into add-project-baseconfig

Undertone0809 committed:
perf: enhance Makefile and documentation for project configuration

- Added `install-docs` target to Makefile for easier documentation setup.
- Renamed `run-docs` to `start-docs` for clarity.
- Expanded documentation on project-based configuration, including a real-world example for standardizing commit messages.
- Improved configuration display in the `show_config` function to include merged effective configuration and better error handling for missing configs.
- Updated test descriptions for clarity and consistency.

Undertone0809 committed:
build: update build actions

Undertone0809 committed:
refactor: rename model_config to model in GcopConfig and main module

Undertone0809 committed:
Merge branch 'main' into add-project-baseconfig

Undertone0809 committed:
docs: enhance README and documentation for GCOP features and configuration

Undertone0809 reviewed (changes requested):
Current situation: If project config is not written with model config, I will give me a default config

Ideally: project config If model config is not written, project model config should be empty, or there is no model field
